### PR TITLE
[BOLT][AArch64] Reduce the number of ADR relaxations

### PR DESCRIPTION
**Pull request by maksfb**
If ADR instruction references the same function, we can skip relaxation even if the function is split, but ADR is in the main fragment.

### CODE REVIEW

File: bolt/test/AArch64/adr-relaxation.s [hunk 1]
```diff
@@ -0,0 +1,31 @@
+## Check that llvm-bolt will unnecessarily relax ADR instruction.
```
Comment (dcci) [inline bolt/test/AArch64/adr-relaxation.s:1]: you mean "will *not* relax" ?

File: bolt/test/AArch64/adr-relaxation.s [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+## ADR below references containing function that is split. But ADR is always
+## in the main fragment, thus there is no need to relax it.
+
+# RUN: llvm-mc -filetype=obj -triple aarch64-unknown-unknown %s -o %t.o
+# RUN: %clang %cflags %t.o -o %t.exe -Wl,-q -static
+# RUN: llvm-bolt %t.exe -o %t.bolt --split-functions --split-strategy=randomN \
+# RUN:   2>&1 | FileCheck %s
+# RUN: llvm-objdump -d --disassemble-symbols=_start %t.bolt | FileCheck %s
+
+# CHECK-NOT: adrp
```
Comment (dcci) [inline bolt/test/AArch64/adr-relaxation.s:11]: can we dump the whole function (or some context?) that way we can make the test more robust.

File: bolt/lib/Passes/ADRRelaxationPass.cpp [hunk 3]
```diff
@@ -56,13 +56,14 @@ void ADRRelaxationPass::runOnFunction(BinaryFunction &BF) {
...
 
-      // Don't relax adr if it points to the same function and it is not split
-      // and BF initial size is < 1MB.
+      // Don't relax ADR if it points to the same function and is in the main
+      // fragment and BF initial size is < 1MB.
       const unsigned OneMB = 0x100000;
       if (BF.getSize() < OneMB) {
         BinaryFunction *TargetBF = BC.getFunctionForSymbol(Symbol);
-        if (TargetBF == &BF && !BF.isSplit())
+        if (TargetBF == &BF && !BB.isSplit())
```
Comment (dcci) [inline bolt/lib/Passes/ADRRelaxationPass.cpp:64]: This looks fine to me. Can we add a negative test? (e.g. one where the func is not in the main fragment to show the adrp is still there?)

**Review by dcci — COMMENTED** [review]
Thanks for working on this. Some comments/questions.

File: bolt/test/AArch64/adr-relaxation.s [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+## ADR below references its containing function that is split. But ADR is always
+## in the main fragment, thus there is no need to relax it.
+	.text
+  .globl _start
+  .type _start, %function
+_start:
+# CHECK: <_start>:
+	.cfi_startproc
+  adr x1, _start
```
Comment (yota9) [inline bolt/test/AArch64/adr-relaxation.s:18]: Indentation

**Review by yota9 — APPROVED** [review]
Thanks!

**Review by dcci — APPROVED** [review]
LGTM

**Pull request merged**